code/modules/goonchat/browserOutput.dm

```diff
@@ -0,0 +1,289 @@
+				log_admin("[key_name(src.owner)] has a cookie from a banned account! (Matched: [found["ckey"]], [found["ip"]], [found["compid"]])")
+
+	cookieSent = TRUE
+
+//Called by js client every 60 seconds
+/datum/chatOutput/proc/ping()
+	return "pong"
+
+//Called by js client on js error
+/datum/chatOutput/proc/debug(error)
+	log_world("\[[time2text(world.realtime, "YYYY-MM-DD hh:mm:ss")]\] Client: [(src.owner.key ? src.owner.key : src.owner)] triggered JS error: [error]")
+
+#ifdef TESTING
+/client/verb/debug_chat()
+	set hidden = TRUE
+	chatOutput.ehjax_send(data = list("firebug" = TRUE))
+#endif
+//Global chat procs
+
+GLOBAL_LIST_EMPTY(bicon_cache)
+
+//Converts an icon to base64. Operates by putting the icon in the iconCache savefile,
+// exporting it as text, and then parsing the base64 from that.
+// (This relies on byond automatically storing icons in savefiles as base64)
+/proc/icon2base64(icon/icon, iconKey = "misc")
+	if (!isicon(icon))
+		return FALSE
+	GLOB.iconCache[iconKey] << icon
+	var/iconData = GLOB.iconCache.ExportText(iconKey)
+	var/list/partial = splittext(iconData, "{")
+	return replacetext(copytext(partial[2], 3, -5), "\n", "")
+
+/proc/bicon(obj)
+	if (!obj)
+		return
+
+	if (isicon(obj))
+		//Icons get pooled constantly, references are no good here.
+		/*if (!bicon_cache["\ref[obj]"]) // Doesn't exist yet, make it.
+			bicon_cache["\ref[obj]"] = icon2base64(obj)
+		return "<img class='icon misc' src='data:image/png;base64,[bicon_cache["\ref[obj]"]]'>"*/
+		return "<img class='icon misc' src='data:image/png;base64,[icon2base64(obj)]'>"
+
+	// Either an atom or somebody fucked up and is gonna get a runtime, which I'm fine with.
+	var/atom/A = obj
+	var/key = "[istype(A.icon, /icon) ? "\ref[A.icon]" : A.icon]:[A.icon_state]"
+	if (!GLOB.bicon_cache[key]) // Doesn't exist, make it.
+		var/icon/I = icon(A.icon, A.icon_state, SOUTH, 1)
+		if (ishuman(obj)) // Shitty workaround for a BYOND issue.
+			var/icon/temp = I
+			I = icon()
+			I.Insert(temp, dir = SOUTH)
+		GLOB.bicon_cache[key] = icon2base64(I, key)
+
+	return "<img class='icon [A.icon_state]' src='data:image/png;base64,[GLOB.bicon_cache[key]]'>"
+
+//Costlier version of bicon() that uses getFlatIcon() to account for overlays, underlays, etc. Use with extreme moderation, ESPECIALLY on mobs.
+/proc/costly_bicon(obj)
+	if (!obj)
+		return
+
+	if (isicon(obj))
+		return bicon(obj)
+
+	var/icon/I = getFlatIcon(obj)
+	return bicon(I)
+
+/proc/to_chat(target, message)
+	if(isnull(target))
+		return
+	//Ok so I did my best but I accept that some calls to this will be for shit like sound and images
+	//It stands that we PROBABLY don't want to output those to the browser output so just handle them here
+	if (istype(message, /image) || istype(message, /sound) || istype(target, /savefile) || !(ismob(target) || islist(target) || istype(target, /client) || istype(target, /datum/log) || target == world))
```

Comment:
On goonstation I have since added a condition to check for target == world and trigger an output loop if so.